main/src/library/BufReader.flix

```diff
@@ -15,78 +15,80 @@
 ///
-/// A call to `read` will read directly from the buffer if bytes are available, and
+/// A call to `read` will read directly from the buffer if items are available, and
 /// will trigger a `read` from the underlying `reader` when the buffer is empty.
 ///
 /// This has the effect of reducing the overall number of `read`s to the underlying
-/// `Reader`.
+/// `Readable`.
 ///
-/// A call to `read` might read more than the requested number of bytes from the
+/// A call to `read` might read more than the requested number of items from the
 /// underlying reader, in order to fill the entire buffer.
 ///
 /// A call to `read` might not read from the underlying `reader` at all, if there are
-/// enough bytes in the buffer.
+/// enough items in the buffer.
 ///
 /// `buffer`: the underlying buffer.
 /// `cursor`: the index in `buffer` of the first element logically considered to be in the buffer.
 /// `size`: the number of elements (starting at `cursor`) logically considered to be in the buffer.
-/// `reader`: the underlying `Reader` being read from.
+/// `reader`: the underlying `Readable` being read from.
 ///
-pub struct BufReader[t: Type, r: Region] {
-    buffer: Array[Int8, r],
+pub struct BufReader[a: Type, t: Type, r: Region] {
+    buffer: Array[a, r],
     mut cursor: Int32,
     mut size: Int32,
     reader: t
 }
 
-instance Reader[BufReader[t, rr]] with Reader[t] {
+instance Readable[BufReader[a, t, rr]] with Readable[t] where Readable.Elm[t] ~ a {
 
-    type Aef = rr + Reader.Aef[t]
+    type Elm = a
 
-    pub def read(buffer: Array[Int8, rb], bufReader: BufReader[t, rr]): Result[IoError, Int32] \ rb + rr + Reader.Aef[t] =
+    type Aef = rr + Readable.Aef[t]
+
+    pub def read(buffer: Array[a, rb], bufReader: BufReader[a, t, rr]): Result[IoError, Int32] \ rb + rr + Readable.Aef[t] =
         BufReader.read(buffer, bufReader)
 
 }
 
 mod BufReader {
 
     ///
-    /// Constructs a `BufReader` wrapping `reader` with a default capacity of 4096 bytes.
+    /// Constructs a `BufReader` wrapping `reader` with a default capacity of 4096 items.
     ///
-    pub def withDefaultCapacity(rc: Region[r], reader: t): BufReader[t, r] \ r =
-        withCapacity(rc, defaultBufferCapacity(), reader)
+    pub def withDefaultCapacity(rc: Region[r], defaultValue: a, reader: t): BufReader[a,t, r] \ r with Readable[t] =
+        withCapacity(rc, defaultBufferCapacity(), defaultValue, reader)
 
     ///
-    /// Constructs a `BufReader` wrapping `reader` with the specified `capacity` in bytes.
+    /// Constructs a `BufReader` wrapping `reader` with the specified `capacity` in items.
     ///
     /// If `capacity <= 0`, the `BufReader` will delegate directly to `reader` without buffering.
     ///
-    pub def withCapacity(rc: Region[r], capacity: Int32, reader: t): BufReader[t, r] \ r =
+    pub def withCapacity(rc: Region[r], capacity: Int32, defaultValue: a, reader: t): BufReader[a, t, r] \ r with Readable[t] =
```

Comment:
The defaultValue doesn't seem ideal, i just need to initialize an empty array

Reply:
I will merge this, but please fix the Array.empty :-)